docs_nnx/guides/randomness.md

```diff
@@ -0,0 +1,245 @@
+---
+jupytext:
+  formats: ipynb,md:myst
+  text_representation:
+    extension: .md
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.13.8
+---
+
+# Randomness
+
+Random state in Flax NNX is radically simplified compared to systems like Haiku/Flax Linen in that it defines "random state as object state". In essence, this means that random state is just another type of state, and is stored in Variables and help by the models themselves. The main characteristic of the RNG system in Flax NNX is that its **explicit**, its **order-based**, and uses **dynamic counters**. This is a bit different from [Flax Linen's RNG system](https://flax.readthedocs.io/en/latest/guides/flax_fundamentals/rng_guide.html) which is (path + order)-based and uses a static counters.
+
+```{code-cell} ipython3
+from flax import nnx
+import jax
+from jax import random, numpy as jnp
+```
+
+## Rngs, RngStream, and RngState
+Flax NNX provides the `nnx.Rngs` type a the main convenice API for managing random state. Following Flax Linen's footsteps, `Rngs` has the ability to create multiple named RNG streams, each with its own state, for the purpose of allowing for tight control over randomness in the context of JAX transforms. Here's a breakdown of the main RNG-related types in Flax NNX:
+
+* **Rngs**: The main user interface. It defines a set of named `RngStream` objects.
+* **RngStream**: A object that can generate a stream of RNG keys. It holds a root `key` and a `count` inside a `RngKey` and `RngCount` Variables respectively. When a new key is generated, the count is incremented.
+* **RngState**: The base type for all RNG-related state.
+  * **RngKey**: Variable type for holding RNG keys, it includes a `tag` attribute containing the name of the stream.
+  * **RngCount**: Variable type for holding RNG counts, it includes a `tag` attribute containing the name of the stream.
+
+To create an `Rngs` object you can simply pass a integer seed or `jax.random.key` instance to any keyword argument of your choice in the constructor. Here's an example:
+
+```{code-cell} ipython3
+rngs = nnx.Rngs(params=0, dropout=random.key(1))
+nnx.display(rngs)
+```
+
+Notice that the `key` and `count` Variables contain the stream name in a `tag` attribute. This is primarily used for filtering as we'll see later.
+
+To generate new keys, you can access one of the streams and use its `__call__` method with no arguments. This will return a new key by using `random.fold_in` with the current `key` and `count`. The `count` is then incremented so that subsequent calls will return new keys.
+
+```{code-cell} ipython3
+params_key = rngs.params()
+dropout_key = rngs.dropout()
+
+nnx.display(rngs)
+```
+
+Note that the `key` attribute does not change when a new keys are generated.
+
+### Standard stream names
+There are only two standard stream names used by Flax NNX, shown in the table below:
```

Comment:
used by Flax NNX -> used by Flax NNX built-in layers

Just to avoid user confusion that other names are not allowed.